Add common blocking implementation details to docs (#5358)
hashicorp/consul: website/source/api/index.html.md

concurrent requests. This adds up to `wait / 16` additional time to the maximum
duration.

### Implementation Details

> While the mechanim is relatively simple to work with, there are a few subtelties

Reviewer: ```suggestion
While the mechanism is relatively simple to work with, there are a few subtleties
```

Reviewer 2: Can we replace subtleties with a more commonly used phrase? Perhaps, "characteristics" or "qualities".

Reviewer: Maybe guidelines?

Author: I guess the intent is that these are non-obvious ways that the naive implementation can fail. "Subtelties" implies that.

I'm not sure any of those other suggestions are quite applicable. Do you think it would be clearer to say: "...there are a few subtle edge cases that must be handled correctly"?

Reviewer 2: Yes, I do like that phrasing better.

> that a robust client must observe in order to not behave badly in edge cases.

Reviewer 2: ```suggestion
that a robust client must observe in order to work correctly, even in edge cases.
```

>  * **Reset the index if it goes backwards**. While indexes in general are 

Reviewer 2: ```suggestion
 * **Reset the index if it decreases below zero**. While indexes in general are 
```

Author: That's not quite the right meaning - it will never be less than zero but might "go backwards" in that you'll get a smaller number than you had before.

Can you think of a better way of saying that? Technically it would be more accurate to say "breaks monotonicity" but that's not better for understanding.

>    monotonically increasing, there are several real-world scenarios in 

Reviewer 2: Can we link to a definition for "monotonically"?

Author: How about: "While indexes in general are monotonically increasing (i.e. they should only ever increase as time passes), "

   which they can go backwards for a given query. Implementations must check 
   to see if a returned index is lower than the previous value, 
   and if it is, should reset index to `0` - effectively restarting their blocking loop. 
   Failure to do so may cause the client to miss future updates for an unbounded 
   time, or to use an invalid index value that causes no blocking and increases 
   load on the servers. Cases where this can occur include:

>    * If a raft snapshot is restored on the servers with older version of the data

Reviewer 2: ```suggestion
   * If a raft snapshot is restored on the servers with older version of the data.
```

>    * KV list operations where an item with the highest index is removed

Reviewer 2: ```suggestion
   * KV list operations where an item with the highest index is removed.
```

>    * A consul upgrade changes the way watches work to optimise them with more 

Reviewer: ```suggestion
   * A Consul upgrade changes the way watches work to optimise them with more 
```

>    * A consul upgrade changes the way watches work to optimise them with more 

Reviewer 2: ```suggestion
   * A consul upgrade changes the way watches work to optimise them with more.
```

   granular indexes.
 * **Sanity check index is greater than zero**. After the initial request (or a
   reset as above) the `X-Consul-Index` returned _should_ always be greater than zero. It
   is a bug in Consul if it is not, however this has happened a few times and can
   still be triggered on some older Consul versions. It's especially bad because it
   causes blocking clients that are not aware to enter a busy loop, using excessive 
   client CPU and causing high load on servers. It is _always_ safe to use an 
   index of `1` to wait for updates when the data being requested doesn't exist
   yet, so clients _should_ sanity check that their index is at least 1 after 
   each blocking response is handled to be sure they actually block on the next 
   request.
 * **Rate limit**. The blocking query mechanism is reasonably efficient when updates 

>  are relatively rate (order of tens of seconds to minutes between updates). In cases 

Reviewer 3: s/rate/rare/

 where a result gets updated very fast however - possibly during an outage or incident 
 with a badly behaved client - blocking query loops degrade into busy loops that 

>  consume execessive client CPU and causing high server load. While it's possible to just add a sleep 

Reviewer 4: ```suggestion
 consume excessive client CPU and causing high server load. While it's possible to just add a sleep 
```

Reviewer 4: `degrade + causing` is mixing tenses. `degrade + cause` uses the same tense

 to every iteration of the loop, this is **not** recommended since it causes update 
 delivery to be delayed in the happy case, and it can exacerbate the problem since 
 it increases the chance that the index has changed on the next request. Clients 
 _should_ instead rate limit the loop so that in the happy case they proceed without 
 waiting, but when values start to churn quickly they degrade into polling at a 
 reasonable rate (say every 15 seconds). Ideally this is done with an algorithm that 
 allows a couple of quick successive deliveries before it starts to limit rate - a 
 [token bucket](https://en.wikipedia.org/wiki/Token_bucket) with burst of 2 is a simple

>  way to acheive this.

Reviewer 4: ```suggestion
 way to achieve this.
```
